lib/src/json_serializable_generator.dart

```diff
@@ -221,14 +227,58 @@ class JsonSerializableGenerator
         // ...resetting `isList` to `false`.
         isList = false;
       }
-    }
 
-    if (!isList && _coreIterableChecker.isAssignableFromType(fieldType)) {
-      // Then we need to add `?.toList()
-      expression += "?.toList()";
+      if (!isList) {
+        // Then we need to add `?.toList()
+        expression += "?.toList()";
+      }
+
+      return expression;
+    } else if (_coreMapChecker.isAssignableFromType(fieldType)) {
+      var args = _getTypeArguments(fieldType, _coreMapChecker);
+      assert(args.length == 2);
+
+      var keyArg = args.first;
+      var valueType = args.last;
+
+      // We're not going to handle converting key types at the moment
+      // So the only safe types for key are dynamic/Object/String
+      var safeKey = keyArg.isDynamic ||
+          keyArg.isObject ||
+          _coreStringChecker.isExactlyType(keyArg);
+
+      var safeValue = false;
+      if (valueType.isDynamic || valueType.isObject) {
+        safeValue = true;
+      } else {
+        safeValue = _stringBoolNumChecker.isAssignableFromType(valueType);
+      }
+
+      if (safeKey) {
+        if (safeValue) {
+          return expression;
+        }
+
+        // convert
+
+        var substitute = "v$depth";
+        var subFieldValue =
+            _fieldToJsonMapValue(substitute, valueType, depth + 1);
+
+        // In this case, we're going to create a new Map with matching reified
+        // types.
+        return "$expression == null ? null :"
+            "new Map<String, dynamic>.fromIterables("
+            "$expression.keys,"
+            "$expression.values.map(($substitute) => $subFieldValue))";
+      }
+    } else if (fieldType.isDynamic ||
+        fieldType.isObject ||
+        _stringBoolNumChecker.isAssignableFromType(fieldType)) {
+      return expression;
     }
 
-    return expression;
+    return "$expression /* unsafe */";
```

Comment:
This went from 46 lines to 80 lines - consider breaking out some composed methods